### PR TITLE
Adding support for XRay Reports API

### PR DESCRIPTION
**Pull request by romelusw**
…/display/JFROG/Xray+REST+API#XrayRESTAPI-REPORTS

- [x] All [tests](https://github.com/jfrog/jfrog-client-go#tests) passed. If this feature is not already covered by the tests, I added new tests.
- [X] This pull request is on the dev branch.
- [X] I used gofmt for formatting the code before submitting the pull request.
-----

### CODE REVIEW

**Comment by romelusw** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by romelusw** [comment]
recheck

File: xray/services/report.go [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+	"github.com/jfrog/jfrog-client-go/utils/errorutils"
+	"net/http"
+)
+
+const (
+	reportsAPI         = "api/v1/reports"
+	vulnerabilitiesAPI = reportsAPI + "/vulnerabilities"
+)
+
+// ReportService defines the Http client and XRay details
```
Comment (eyalbe4) [inline xray/services/report.go:20]: Let's change "XRay" to "Xray" (let's make sure to fix in other places in this file as well).

**Review by eyalbe4 — CHANGES_REQUESTED** [review]
Let's document the new APIs in the README as well. I also suggest we explain how the new APIs should be used together. For example, how to retrieve the report ID when invoking the ```ReportDetails```, what is ```ReportDetails```  used for, and so on.  This information should also be available as part of the APIs description in ```manager.go``` and maybe also in ```report.go```.

It is also good if the README includes an example, to make it easier to use the API.

File: xray/services/scan.go [hunk 2]
```diff
@@ -226,11 +226,11 @@ type ImpactPathNode struct {
...
-	Id           string `json:"cve,omitempty"`
-	CvssV2Score  string `json:"cvss_v2_score,omitempty"`
-	CvssV2Vector string `json:"cvss_v2_vector,omitempty"`
-	CvssV3Score  string `json:"cvss_v3_score,omitempty"`
-	CvssV3Vector string `json:"cvss_v3_vector,omitempty"`
+	Id           string  `json:"cve,omitempty"`
+	CvssV2Score  float64 `json:"cvss_v2_score,omitempty"`
+	CvssV2Vector string  `json:"cvss_v2_vector,omitempty"`
+	CvssV3Score  float64 `json:"cvss_v3_score,omitempty"`
+	CvssV3Vector string  `json:"cvss_v3_vector,omitempty"`
```
Comment (eyalbe4) [inline xray/services/scan.go:233]: I suggest we avoid modifying this struct, since it is used for other purposes, and this change seem to be breaking them. If needed, we can add a new struct in ```report.go```.

Comment (romelusw) [reply xray/services/scan.go:233]: The API response for CVSScores are actually numerics, not string types so I presumed this was an oversight?

File: xray/services/report.go (hunk 1 above)
Comment (romelusw) [reply xray/services/report.go:20]: I have updated the readme with sample usage.

**Comment by romelusw** [comment]
recheck

**Comment by romelusw** [comment]
I have read the CLA Document and I hereby sign the CLA

File: xray/services/scan.go (hunk 2 above)
Comment (romelusw) [reply xray/services/scan.go:233]: @eyalbe4 any update?

File: utils/vcsdetails.go [hunk 3]
```diff
@@ -13,10 +13,10 @@ import (
 type (
 	VcsCache struct {
 		// Key - Path to the .git directory.
-		// Value - Reference to a struct, storing the URL and revision.
+		// Number - Reference to a struct, storing the URL and revision.
```
Comment (eyalbe4) [inline utils/vcsdetails.go:16]: Let'e revert this change, because "value", indicates the map's value. "Number" seems wrong.

File: utils/vcsdetails.go [hunk 4]
```diff
@@ -13,10 +13,10 @@ import (
...
 		vcsRootDir sync.Map
 		// Key - Path to a directory.
-		// Value - Reference to a struct, storing the URL and revision from the upstream .git. Can also include nil, if there's no upstream .git.
+		// Number - Reference to a struct, storing the URL and revision from the upstream .git. Can also include nil, if there's no upstream .git.
```
Comment (eyalbe4) [inline utils/vcsdetails.go:19]: Let'e revert this change, because "value", indicates the map's value. "Number" seems wrong.

File: xray/services/report.go [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+type CvssScore struct {
+	MinScore float32 `json:"min_score,omitempty"`
+	MaxScore float32 `json:"max_score,omitempty"`
+}
+
+type Resource struct {
+	IncludePathPatterns []string     `json:"include_path_patterns,omitempty"`
+	Repositories        []Repository `json:"repositories,omitempty"`
+}
+type Repository struct {
```
Comment (eyalbe4) [inline xray/services/report.go:101]: Missing new-line before this structs's definition.

File: xray/services/report.go [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+	}
+
+	resp, body, err := rs.client.SendPost(url, content, &httpClientsDetails)
+	if err = errorutils.CheckResponseStatus(resp, http.StatusOK); err != nil {
+		return &retVal, errorutils.CheckError(errorutils.GenerateResponseError(resp.Status, clientutils.IndentJson(body)))
+	}
+
+	err = json.Unmarshal(body, &retVal)
+	if err != nil {
+		return &retVal, errors.New("failed unmarshalling report response")
```
Comment (eyalbe4) [inline xray/services/report.go:135]: Missing CheckError here. This line should be changed to - 
```
return &retVal, errorutils.CheckError(err)
```

File: xray/services/report.go [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+
+	url := fmt.Sprintf("%s/%s/%s", rs.XrayDetails.GetUrl(), reportsAPI, reportId)
+	resp, body, _, err := rs.client.SendGet(url, true, &httpClientsDetails)
+	if err = errorutils.CheckResponseStatus(resp, http.StatusOK); err != nil {
+		return &retVal, errorutils.CheckError(errorutils.GenerateResponseError(resp.Status, clientutils.IndentJson(body)))
+	}
+
+	err = json.Unmarshal(body, &retVal)
+	if err != nil {
+		return &retVal, errors.New("failed unmarshalling report details " + reportId)
```
Comment (eyalbe4) [inline xray/services/report.go:155]: Missing CheckError here. This line should be changed to -
```
return &retVal, errorutils.CheckError(err)
```

File: xray/services/report.go [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+	url := fmt.Sprintf("%s/%s/%s?direction=%s&page_num=%d&num_of_rows=%d&order_by=%s",
+		rs.XrayDetails.GetUrl(), vulnerabilitiesAPI, request.ReportId, request.Direction, request.PageNum, request.NumRows, request.OrderBy)
+	resp, body, err := rs.client.SendPost(url, nil, &httpClientsDetails)
+	if err = errorutils.CheckResponseStatus(resp, http.StatusOK); err != nil {
+		return &retVal, errorutils.CheckError(errorutils.GenerateResponseError(resp.Status, clientutils.IndentJson(body)))
+	}
+
+	err = json.Unmarshal(body, &retVal)
+	if err != nil {
+		return &retVal, errors.New("failed unmarshalling content for report " + request.ReportId)
```
Comment (eyalbe4) [inline xray/services/report.go:176]: Missing CheckError here. This line should be changed to -
```
return &retVal, errorutils.CheckError(err)
```

File: xray/services/report.go [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+	}
+	return &retVal, nil
```
Comment (eyalbe4) [inline xray/services/report.go:178]: The above if statement can be removed, and code can be simplified to - 
```
	err = json.Unmarshal(body, &retVal)
	return &retVal, errorutils.CheckError(err)
```

File: xray/services/scan.go (hunk 2 above)
Comment (eyalbe4) [reply xray/services/scan.go:233]: Let's leave them as strings for now. If this causes issues for the new reporting functionality, please let me know, and we'll find a solution for this.
You change simply breaks jfrog-cli-core, which uses this as a dependency.

File: README.md [hunk 10]
```diff
@@ -1470,6 +1474,51 @@ scanId, err := xrayManager.ScanGraph(graphScanParams)
...
+    IncludePathPatterns: []string{ "/example-sub-dir/**" },
+    Repositories: []services.Repository{
+      {
+        Name: "example-repository",
+      },
+    },
+  },
+}
+
+// The reportRequestResponse will contain the report id to use in subsequent requests
```
Comment (eyalbe4) [inline README.md:1496]: id --> ID

File: README.md [hunk 11]
```diff
@@ -1470,6 +1474,51 @@ scanId, err := xrayManager.ScanGraph(graphScanParams)
...
+reportRequestResponse, err := xrayManager.GenerateVulnerabilitiesReport(reportRequest)
+```
+
+#### Get Vulnerabilities Report Details
+```go
+reportDetails, err := xrayManager.ReportDetails(reportId)
```
Comment (eyalbe4) [inline README.md:1502]: Let's add the following comment above this line:
// The reportId argument value is returned as part of the xrayManager.GenerateVulnerabilitiesReport API response.

File: README.md [hunk 12]
```diff
@@ -1470,6 +1474,51 @@ scanId, err := xrayManager.ScanGraph(graphScanParams)
...
+```
+
+#### Get Vulnerabilities Report Content
+```go
+reportContentRequest := services.ReportContentRequestParams{
```
Comment (eyalbe4) [inline README.md:1507]: Let's add the following comment above this line:
// The ReportId value is returned as part of the xrayManager.GenerateVulnerabilitiesReport API response.

File: README.md [hunk 13]
```diff
@@ -1470,6 +1474,51 @@ scanId, err := xrayManager.ScanGraph(graphScanParams)
...
+  PageNum:   0,
+  NumRows:   0,
+  OrderBy:   "severity",
+}
+reportContent, err := xrayManager.ReportContent(reportContentRequest)
+```
+
+#### Delete Vulnerabilities Report
+```go
+err := xrayManager.DeleteReport(reportId)
```
Comment (eyalbe4) [inline README.md:1519]: Let's add the following comment above this line:
// The reportId argument value is returned as part of the xrayManager.GenerateVulnerabilitiesReport API response.

**Review by eyalbe4 — CHANGES_REQUESTED** [review]
Let's also add an integration test, which generates a report, fetches it, and the deletes it, using all the new APIs added.
You can see the ```TestXrayWatch``` test a reference for how this is done, or any of the other tests under the ```tests``` package.

**Comment by romelusw** [comment]
@eyalbe4, @yahavi can you please approve this

**Review by eyalbe4 — CHANGES_REQUESTED** [review]
Thanks you so much for this contribution! Looks great!
Just a minor suggestion - see inline.
We'd happy to merge this following your change.
